lib/ash/type/string.ex

```diff
@@ -34,39 +47,71 @@ defmodule Ash.Type.String do
   def apply_constraints(nil, _), do: :ok
 
   def apply_constraints(value, constraints) do
+    opts = NimbleOptions.validate!(constraints, @constraints)
+
+    trim_value? = opts[:trim?]
+    allow_empty? = opts[:allow_empty?]
+
+    trimmed_value = String.trim(value)
```

Comment:
One thing to note here: trimming a large string could be expensive, and here we are trimming it at all times. But if `trim?` is `false` and `allow_empty` is `true`, we never actually use the `trimmed_value`. So we should choose a control flow pattern here that doesn't trim the string 100% of the time.